src/Tutor.Infrastructure/Database/EventStore/DefaultEventSerializer/AllowedTypesDiscriminatorConvention.cs

```diff
@@ -19,30 +19,26 @@ public class AllowedTypesDiscriminatorConvention<T> : IDiscriminatorConvention w
 
         public ReadOnlySpan<byte> MemberName => _memberName.Span;
 
-        public AllowedTypesDiscriminatorConvention(JsonSerializerOptions options, IDictionary<Type, T> allowedTypes)
+        public AllowedTypesDiscriminatorConvention(JsonSerializerOptions options, IImmutableDictionary<Type, T> allowedTypes)
             : this(options, allowedTypes, "$type")
         {
         }
 
-        public AllowedTypesDiscriminatorConvention(JsonSerializerOptions options, IDictionary<Type, T> allowedTypes, string memberName)
+        public AllowedTypesDiscriminatorConvention(JsonSerializerOptions options, IImmutableDictionary<Type, T> allowedTypes, string memberName)
```

Comment:
@Luburic When switching to using immutable dictionaries instead of regular ones I ran into a bug that I think is actually an issue with Dahomey.Json that happens in the following circumstances:
1. A hierarchy contains concrete class A, as well as concrete class B which inherits from A. 
2. When registering classes from the hierarchy in the DiscriminatorConventionRegistry, class B is registered before class A. 

If this happens, registering class A won't actually do anything, so serialization of instances of class A will fail. 

I currently fixed this in the AllowedTypesDiscriminatorConvention by actually performing the work necessary to set up the convention in the constructor instead, including the parts which I believe Dahomey.Json expects to happen while registering types instead. But this patch is a bit odd and might be kind of weird to maintain long term. 

Should we maybe look for another Json library to replace Dahomey.Json? Or will situations like class A and class B ever even happen in event hierarchies? I don't think the current CCaDET hierarchy has any actual case like that (only the test classes for serialization tests do).

Reply:
I think we will always work with the "leaf" children of the hierarchy. However, this is quite a sensitive situation that should be recorded in a comment. Add a few lines of explanation next to the sensitive code and ensure you mention that it is specific to Dahomey.Json.

There are some indications that System.Text.Json might support polymorphic (de)serialization in .NET 7 so I would wait to see if that is the case before looking for another alternative to Dahomey.

Reply:
Added explanatory comments to AllowedTypesDiscriminatorConvention in f5740ce62e3d0b1c4f2f93a908cfa02cb3fbf66a.